### PR TITLE
inject dataset version with AbstractQueryPlanDatasetBuilder

### PR DESCRIPTION
**Pull request by pawel-big-lebowski**
try to use that

### CODE REVIEW

File: integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/DatasetBuilderFactory.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package io.openlineage.spark.agent.lifecycle;
+
+import io.openlineage.client.OpenLineage;
+import io.openlineage.spark.api.OpenLineageContext;
+import java.util.Collection;
+import java.util.List;
+import scala.PartialFunction;
+
+public interface DatasetBuilderFactory {
```
Comment (collado-mike) [inline integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/DatasetBuilderFactory.java:9]: Add some javadoc explaining the need for this interface

File: integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/LogicalPlanSerializer.java [hunk 2]
```diff
@@ -58,7 +59,11 @@ public String serialize(LogicalPlan x) {
     try {
       return mapper.writeValueAsString(x);
     } catch (Throwable e) {
-      return "Unable to serialize: " + e.getMessage();
+      try {
+        return mapper.writeValueAsString("Unable to serialize: " + e.getMessage());
```
Comment (collado-mike) [inline integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/LogicalPlanSerializer.java:63]: 🙄

File: integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/OpenLineageRunEventBuilder.java [hunk 3]
```diff
@@ -452,16 +452,24 @@ private RunEvent populateRun(
...
+            event ->
+                builders.stream()
+                    .filter(
+                        builder -> {
+                          try {
+                            log.info(
+                                "{} isDefinedAT {} returned {}",
+                                builder.getClass(),
+                                event.getClass(),
+                                builder.isDefinedAt(event));
```
Comment (collado-mike) [inline integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/OpenLineageRunEventBuilder.java:466]: At best, this is a debug statement, but I don't really know that it should be here at all

File: integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/plan/AppendDataDatasetBuilder.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        .filter(b -> b.getClass().getName().equals("DataSourceV2RelationOutputDatasetBuilder"))
+        .flatMap(b -> b.apply(logicalPlan).stream())
+        .collect(Collectors.toList());
+
+    // FIXME: need to do something
+    // NEED TO FIX IT;
+    //    return PlanUtils.applyFirst(
+    //        context.getOutputDatasetBuilders(), (LogicalPlan) ((AppendData) x).table()
+    //    );
+  }
```
Comment (collado-mike) [inline integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/plan/AppendDataDatasetBuilder.java:50]: The `AbstractQueryPlanDatasetBuilder#delegate` method should be what you're looking for here.

File: integration/spark/src/main/common/java/io/openlineage/spark/api/AbstractQueryPlanDatasetBuilder.java [hunk 5]
```diff
@@ -81,14 +84,16 @@ public boolean isDefinedAt(LogicalPlan x) {
...
+  public boolean isDefinedAtLogicalPlan(LogicalPlan logicalPlan) {
     Type genericSuperclass = getClass().getGenericSuperclass();
+    log.info("Checking isDefinedAtLogicalPlan for {}", genericSuperclass.getTypeName());
     if (!(genericSuperclass instanceof ParameterizedType)) {
       return false;
     }
     Type[] typeArgs = ((ParameterizedType) genericSuperclass).getActualTypeArguments();
     if (typeArgs != null && typeArgs.length > 1) {
       Type arg = typeArgs[1];
+      log.info("check if is isAssignableFrom {}", logicalPlan.getClass());
```
Comment (collado-mike) [inline integration/spark/src/main/common/java/io/openlineage/spark/api/AbstractQueryPlanDatasetBuilder.java:96]: debug?

File: integration/spark/src/main/common/java/io/openlineage/spark/api/AbstractQueryPlanInputDatasetBuilder.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    super(context);
+  }
+
+  @Override
+  public boolean isDefinedAt(SparkListenerEvent event) {
+    // FIXME: missing test for this
+
+    return event instanceof SparkListenerJobStart
+        || event instanceof SparkListenerSQLExecutionStart;
+  }
```
Comment (collado-mike) [inline integration/spark/src/main/common/java/io/openlineage/spark/api/AbstractQueryPlanInputDatasetBuilder.java:22]: What's the reasoning for this? (should go in the javadoc)

File: integration/spark/src/main/common/java/io/openlineage/spark/api/AbstractQueryPlanOutputDatasetBuilder.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+  public AbstractQueryPlanOutputDatasetBuilder(OpenLineageContext context) {
+    super(context);
+  }
+
+  @Override
+  public boolean isDefinedAt(SparkListenerEvent event) {
+    // FIXME: missing test for this
+    return event instanceof SparkListenerJobEnd || event instanceof SparkListenerSQLExecutionEnd;
+  }
```
Comment (collado-mike) [inline integration/spark/src/main/common/java/io/openlineage/spark/api/AbstractQueryPlanOutputDatasetBuilder.java:20]: Reasoning?

File: integration/spark/src/main/common/java/io/openlineage/spark/api/QueryPlanVisitor.java [hunk 8]
```diff
@@ -103,6 +106,10 @@ public boolean isDefinedAt(LogicalPlan x) {
     return false;
   }
 
+  public void setTriggeringEvent(SparkListenerEvent triggeringEvent) {
+    this.triggeringEvent = triggeringEvent;
+  }
+
```
Comment (collado-mike) [inline integration/spark/src/main/common/java/io/openlineage/spark/api/QueryPlanVisitor.java:112]: We don't need this anymore, right?

File: integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/LogicalPlanSerializer.java (hunk 2 above)
Comment (pawel-big-lebowski) [reply integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/LogicalPlanSerializer.java:63]: Nasty bug. More details can be found in this PR: https://github.com/OpenLineage/OpenLineage/pull/581
This should be out from this PR.

**Pull request closed without merging**